File: day1/day1_1.py
import bisect
import sys

# Check if a filename was provided
if len(sys.argv) < 2:
    print("Usage: python script.py <filename>")
    sys.exit(1)

# Read the filename from the command line
filename = sys.argv[1]

# Initialize empty lists to store the numbers
llist = []
rlist = []
dlist = []
sim_score = []

distance_sum = 0

# Open the file in read mode
with open(filename, 'r') as file:
    for line in file:
        # Split the line into two parts and convert to numbers
        parts = line.split()
        if len(parts) == 2:  # Ensure the line has exactly two parts
            left, right = map(int, parts)  # Convert to int
            print(f'Left is {left}, right is {right}')
            # llist.append(left))
            bisect.insort(llist, left)
            # rlist.append(right))
            bisect.insort(rlist, right)

print(f'Left list ordered is {llist}.\n Right list ordered is {rlist}.')

for l, r in zip(llist, rlist):
    distance_sum = distance_sum + abs(l - r)
    dlist.append(abs(l - r))

print(f"The distance of the lists is {distance_sum}.")


for i, l in enumerate(llist):
    sim_score.append(0)
    for r in rlist:
        if l == r:
            sim_score[i] = sim_score[i] + 1
    sim_score[i] = sim_score[i] * l

print(f"The similarity score list is: {sim_score}. The similarity score is {sum(sim_score)}.")

# Output the results
# print("Left List:", llist)
# print("Right List:", rlist)

File: day1/day1_1.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() -> io::Result<()> {
    // Get command-line arguments
    let args: Vec<String> = env::args().collect();


    let mut dlist: Vec<i32> = Vec::new();
    let mut slist: Vec<i32> = Vec::new();

    // Check if a filename was provided
    if args.len() < 2 {
        eprintln!("Usage: {} <filename>", args[0]);
        return Ok(()); // Gracefully exit
    }

    let file_path = &args[1]; // The first argument after the program name is the filename


    // Call the function to read the file
    /* if let Err(e) = read_lines(file_path) {
        eprintln!("Error reading file: {}", e);
    } */

    
    let (mut llist, mut rlist) = read_lines(file_path)?;
    llist.sort();
    println!("{:?}", llist);
    rlist.sort();
    println!("{:?}", rlist);

    for (l, r) in llist.iter().zip(rlist.iter()) {
        dlist.push((l - r).abs())
    }
    println!("{:?}", dlist);
    let dist_sum: i32 = dlist.iter().sum();
    println!("The sum of the distance vector is: {}", dist_sum);

    for (index, l) in llist.iter().enumerate() {
        slist.push(0);

        for r in &rlist {
            if l == r {
                slist[index] += 1;
            }
        }

        slist[index] *= l
        
    }
    println!("The similarity vector is: {:?}, the similarity score is {}.", slist, slist.iter().sum::<i32>());

    Ok(())

}

fn read_lines<P>(filename: P) -> Result<(Vec<i32>, Vec<i32>), io::Error>
where
    P: AsRef<Path>,
{
    let mut llist: Vec<i32> = Vec::new();
    let mut rlist: Vec<i32> = Vec::new();
    let mut nums: Vec<i32> = Vec::new();

    // Open the file
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);

    // Iterate over the lines
    for (index, line) in reader.lines().enumerate() {
        let line = line?; // Unwrap the Result to get the line content
        nums = line.split_whitespace().map(|s| s.parse::<i32>().unwrap()) // parse each part as i32
        .collect();

        println!("Line {}: {} and {}", index + 1, nums[0], nums[1]);
        llist.push(nums[0]);
        rlist.push(nums[1]);
    }

    Ok((llist, rlist))
}
